### PR TITLE
Split libcxx into two parts

### PR DESCRIPTION
**Pull request by gagern**
This is for ticket #2545, an attempt to reduce resulting file size for simple C++ programs.

The I/O subsystem, including locale management, accounts for a large portion of the resulting file size.  On the other hand, many projects will make little to no use of console text I/O or stringstream formatting.  Those projects can now benefit from the fact that they can omit one of the two parts, thereby massively decreasing the size of the generated code.

I checked that according to `ld -r` the files contained in the “slim” part had no unresolved dependencies on the “ios” part, so the cut is assumed to be clean.

### CODE REVIEW

**Comment by gagern** [comment]
This may take some more work, I fear. It's still too easy to end up in a situation where the ios part gets loaded. One cause was symbols which were undefined in libcxx_ios itself. The recent commit should address this. Right now I'm hunting after a symbol of the form `__vector_base_common<__b>::__throw_length_error`. When I run `llvm-nm` manually, it appears in `libcxx_ios.bc` with `W` (meaning weak?), in `libcxx_slim` as `U` for undefined, and in my input with `t` as text. But when `Building.llvm_nm` investigates a file of the same name (in the same dir?), it reports the symbol as `U` undefined and thus links in `libcxx_ios`. Strange!

**Comment by gagern** [comment]
Sorry, my mistake: I was creating a `.bc` file during object compilation, and then using `--save-bc` with that same name during linking. Yes, the symbol is indeed undefined. But why? It's a template instantiation and included in the header. Shouldn't that include the code in each module which makes use of that template? Here is some code to reproduce this:

```
#include <vector>
int main(int argc, char** argv) {
  std::vector<void*>* v =
    reinterpret_cast<std::vector<void*>*>(argv);
  v->push_back(nullptr);
  return 0;
}
```

**Comment by waywardmonkeys** [comment]
It might well be worth discussing this with the ``libcxx`` upstream as well. They have a variety of things that generate fairly large code and it might be good for everyone if they improved that.

(Just wait until you see what ``std::merge_sort`` expands to some day...)

File: tools/system_libs.py [hunk 1]
```diff
@@ -33,11 +33,14 @@ def calculate(temp_files, in_temp, stdout_, stderr_, forced=[]):
...
-    symbols = map(lambda line: line.strip().split(' ')[1], open(path).readlines())
-    if exclude:
-      symbols = filter(lambda symbol: symbol not in exclude, symbols)
-    return set(symbols)
+  def read_symbols(path, exclude=()):
+    symbols = set()
+    with open(path, 'r') as file:
+      for line in file:
+        mark, symbol = line.strip().split(' ')
+        if mark != 'U' and symbol not in exclude:
```
Comment (waywardmonkeys) [inline tools/system_libs.py]: This isn't needed. You can just pass the right args to ``llvm-nm`` when generating the symbols files.  Without verifying it, I think the flags to pass were ``-defined-only -extern-only``.

**Comment by waywardmonkeys** [comment]
As for your undefined symbol from ``vector`` ... are you compiling with exceptions enabled or disabled?

**Comment by kripken** [comment]
It happens with `-O2`, which is without exceptions. `-s VERBOSE=1` says the symbol is `_ZNKSt3__120__vector_base_commonILb1EE20__throw_length_errorEv` aka `std::__1::__vector_base_common<true>::__throw_length_error() const`.

That's just a function call, it is still emitted when `-fno-exceptions` is passed. So we can't avoid it being in the code, and it brings in libc++.

**Comment by gagern** [comment]
Makes me wonder how *that* function was compiled. Looking at the sources, it behaves differently depending on whether you compile with or without exceptions. Which means it doesn't count how you compile your current file, but how `system_libs.py` compiled its library files. Do we need two copies of the system libs, depending on whether or not we use exceptions?

By the way, I guess I'd aim for lazy initialization at some level or other. Perhaps in the `ios_base.getloc()` method. I know lazy initialization can be a real pain for asynchroneous code, but since we can't have that with JavaScript, we can make use of that, simply set some global flag to `false` initially and only create the locale object if it is actually needed. At least if the default locale hasn't ben changed since startup.

By the way, there are some `pthread` symbols in the code as well. Perhaps we should make extra sure that all the code will consider itself single-threaded, so it can make full use of that for optimizations. Or do you actually model threads using workes? I can hardly imagine that, since there is no shared memory…

@kripken: I agree that `__throw_length_error` is just a function call and has to be in the code, but do you know why it's satisfied from the library, and not from the current compilation unit, since it's a template instantiation? Does the compiler at that point know which symbols the library is going to provide? If so, what list does it have? Should we give it a shorter list somehow?

**Comment by kripken** [comment]
I don't know. I did look a bit more closely at the chain here, and it looks like creating that exception creates an object with a vtable and/or RTTI, and that is a bunch of global variables linking to miscellaneous dynamic cast stuff, and I suppose in there somewhere is code using more libc++ code.

Maybe we do need 2 libc++ builds, one with and one without exceptions...

**Comment by kripken** [comment]
Yeah, that seems like what we need. It might even be a bigger deal in fact - as right now even basic `vector` usage seems to pull in locale code. We should probably do that first, before trying the splitting approach, hopefully it won't be needed afterwards.

Thoughts?

**Comment by kripken** [comment]
Discussing this with @sunfishcode , another thought was that likely this would be fixed by building libc++ as a `.a` library. Then `iostream.cpp` wouldn't be pulled in specifically unless it was needed. That might be better than a single split point.

**Comment by gagern** [comment]
That `.a` library approach sounds interesting, though I'd have to see it in action before I trust that it does the right thing regarding `ios_base::Init::Init()`. And it still would initialize the locale eagerly, so most of the code would be included the moment someone uses any stream at all.

Basic `vector` usage only pulls in locale support because it pulls in the library with its global constructor call. There is nothing specific to `vector` here except for the fact that the ios part of the library apparently uses `vector`, which is fine, and the fact that for some reason that one vector symbol is actually resolved using the ios library instead of the current translation unit, which I still don't understand.

**Comment by kripken** [comment]
@sunfishcode explained that part to me, there is some C++ feature which libc++ uses to force something defined in a header to *not* be emitted inline or in each compilation module, but to just emit an extern. Then libc++ defines it in the library itself. Not sure why they do that.

**Comment by kripken** [comment]
Just using libc++ built without exceptions doesn't help a lot. The `throw_length_error` method has already caused all of libc++ to be linked in, even if now without exceptions it only aborts, and doesn't throw an exception. I'll test `.a` creation now.

**Comment by kripken** [comment]
Ok, the [minimal_libcxx](https://github.com/kripken/emscripten/commits/minimal_libcxx) branch uses a .a, and has a noexcept version as well. This makes simple vector usage much smaller. Running tests now.

**Comment by gagern** [comment]
The C++11 feature to avoid including a template instantiation would imho be an `extern template` line somewhere. I hadn't found any of those around, but a closer look just now showed me that the `_LIBCPP_EXTERN_TEMPLATE` and `_LIBCPP_EXTERN_TEMPLATE2` macros have just that definition. The `<vector>` header contains an extern template declaration for `__vector_base_common<true>`. And the implementation for it is indeed in `locale.cc`. Whoever put it there… So if we want to make the split library more efficient, we could move that line someplace else. I think we should do that for the `.a` approach as well, since otherwise we'll have `locale.cpp` in all our code again, although it might be better since a lot of code from `locale.cpp` by itself might be considered dead without the global initializer.

**Comment by kripken** [comment]
I think I see locale code not being pulled in, at least not for vector. What's another testcase we want to ensure is small here?

**Comment by gagern** [comment]
Thanks for the work here! I fear number to string conversion using stringstreams would be too much to hope for. I'd say `sort` and `map` should be small, and the `shared_ptr` which motivated the split in the first place.

**Comment by kripken** [comment]
Ok. We can make some concrete testcases and try them on that branch, maybe add tests from them if they work (there is one for vector already).

**Comment by kripken** [comment]
Merged to incoming as the tests look fine so far locally, and I'm pretty we'll need this stuff (libc++ without exceptions, and .a so we pull in individual parts of libc++ as needed) even if we do still want to do a manual split.

**Comment by gagern** [comment]
I don't think we'll need a manual split any more. We might need some lazy initialization of some stuff one day, but that's a different matter.

**Comment by waywardmonkeys** [comment]
@gagern If you care about generated code size, you may want to re-think your usage of ``std::sort`` and friends. At least some of the sorting code blows up into something very very large.

**Comment by waywardmonkeys** [comment]
@gagern And you may want to bring up the issue of the ``__vector_base_common<true>`` and ``locale.cc`` upstream ... they're receptive people.

**Comment by kripken** [comment]
Ok, should we close this then? Or is there more stuff we want to investigate and/or do?

**Comment by gagern** [comment]
Let's close this. Thanks for your work.

**Comment by kripken** [comment]
Thanks for the submitting this, it's great we finally looked into these issues and made some progress.

**Pull request closed without merging**